### ISSUE TITLE
Something fishy with ASMu3D projections

### ISSUE THREAD

**Issue by VikingScientist** [OPM/IFEM#197]
Running [3D multipatch adaptive elasticity test "disc-stair-adap"](https://github.com/OPM/IFEM-Elasticity/blob/master/Linear/Test/LR/disc-stair-adap.xinp) and changing `<maxstep>` in `<adaptive>` to 6 makes the test crash in the final adaptive iteration. This segmentation fault happens during continuous global L2-projection of secondary solutions.

**Comment by akva2** [comment]
adding some info so it's stored somewhere;

what happens is that we have no mesh rectangle which is constant v=0 after refinement, ref
```
Refining 21 basis functions with errors in range [0.522302,0.627715] 1% of all basis functions
before refine : [0, 0] x [0, 1] x [0, 1] (3)
[0.5, 0.5] x [0, 1] x [0, 1] (1)
[1, 1] x [0, 1] x [0, 1] (3)
[0, 1] x [0, 0] x [0, 1] (3)
[0, 1] x [0.5, 0.5] x [0, 1] (1)
[0, 1] x [1, 1] x [0, 1] (3)
[0, 1] x [0, 1] x [0, 0] (3)
[0, 1] x [0, 1] x [0.5, 0.5] (1)
[0, 1] x [0, 1] x [1, 1] (3)
[0.25, 0.25] x [0, 1] x [0, 1] (1)
[0.75, 0.75] x [0, 1] x [0, 1] (1)
[0, 1] x [0.25, 0.25] x [0, 1] (1)
[0, 1] x [0.75, 0.75] x [0, 1] (1)
[0, 1] x [0, 1] x [0.25, 0.25] (1)
[0, 1] x [0, 1] x [0.75, 0.75] (1)
[0, 1] x [0.125, 0.125] x [0, 0.75] (1)
[0, 1] x [0.375, 0.375] x [0, 0.75] (1)
[0, 1] x [0.625, 0.625] x [0, 0.75] (1)
[0, 1] x [0, 0.75] x [0.125, 0.125] (1)
[0, 1] x [0, 0.75] x [0.375, 0.375] (1)
[0, 1] x [0, 0.75] x [0.625, 0.625] (1)
[0.125, 0.125] x [0, 0.75] x [0, 1] (1)
[0.375, 0.375] x [0, 0.75] x [0, 1] (1)
[0.625, 0.625] x [0, 0.75] x [0, 1] (1)
[0.875, 0.875] x [0, 0.75] x [0, 1] (1)
[0, 1] x [0.125, 0.125] x [0, 1] (1)
[0, 1] x [0.375, 0.375] x [0, 1] (1)
[0, 1] x [0.625, 0.625] x [0, 1] (1)
[0, 1] x [0, 0.75] x [0.875, 0.875] (1)
[0, 1] x [0.875, 0.875] x [0, 0.75] (1)
[0, 1] x [0, 1] x [0.125, 0.125] (1)
[0, 1] x [0, 1] x [0.375, 0.375] (1)
[0, 1] x [0, 1] x [0.625, 0.625] (1)
[0.125, 0.125] x [0, 1] x [0, 1] (1)
[0.375, 0.375] x [0, 1] x [0, 1] (1)
[0.625, 0.625] x [0, 1] x [0, 1] (1)
[0.875, 0.875] x [0, 1] x [0, 1] (1)
[0, 1] x [0.875, 0.875] x [0, 1] (1)
[0, 1] x [0, 1] x [0.875, 0.875] (1)
[0.0625, 0.0625] x [0, 0.375] x [0, 0.375] (1)
[0.1875, 0.1875] x [0, 0.375] x [0, 0.375] (1)
[0.3125, 0.3125] x [0, 0.375] x [0, 0.375] (1)
[0, 0.375] x [0.0625, 0.0625] x [0, 0.375] (1)
[0, 0.375] x [0.1875, 0.1875] x [0, 0.375] (1)
[0, 0.375] x [0.3125, 0.3125] x [0, 0.375] (1)
[0, 0.375] x [0, 0.375] x [0.0625, 0.0625] (1)
[0, 0.375] x [0, 0.375] x [0.1875, 0.1875] (1)
[0, 0.375] x [0, 0.375] x [0.3125, 0.3125] (1)
[0.4375, 0.4375] x [0, 0.375] x [0.625, 1] (1)
[0.5625, 0.5625] x [0, 0.375] x [0.625, 1] (1)
[0, 0.625] x [0.0625, 0.0625] x [0.625, 1] (1)
[0, 0.625] x [0.1875, 0.1875] x [0.625, 1] (1)
[0, 0.625] x [0.3125, 0.3125] x [0.625, 1] (1)
[0, 0.625] x [0, 0.375] x [0.6875, 0.6875] (1)
[0, 0.625] x [0, 0.375] x [0.8125, 0.8125] (1)
[0, 0.625] x [0, 0.375] x [0.9375, 0.9375] (1)
[0.0625, 0.0625] x [0, 0.375] x [0.5, 1] (1)
[0.1875, 0.1875] x [0, 0.375] x [0.5, 1] (1)
[0.3125, 0.3125] x [0, 0.375] x [0.5, 1] (1)
[0, 0.375] x [0.0625, 0.0625] x [0.5, 1] (1)
[0, 0.375] x [0.1875, 0.1875] x [0.5, 1] (1)
[0, 0.375] x [0.3125, 0.3125] x [0.5, 1] (1)
[0, 0.375] x [0, 0.375] x [0.5625, 0.5625] (1)
after refine : [0, 0] x [0, 1] x [0, 1] (3)
[0, 1] x [0.5, 0.5] x [0, 1] (1)
[0, 1] x [1, 1] x [0, 1] (3)
[0, 1] x [0, 1] x [0, 0] (3)
[0, 1] x [0, 1] x [0.5, 0.5] (1)
[0, 1] x [0, 1] x [1, 1] (3)
[0.25, 0.25] x [0, 1] x [0, 1] (1)
[0.75, 0.75] x [0, 1] x [0, 1] (1)
[0, 1] x [0.25, 0.25] x [0, 1] (1)
[0, 1] x [0.75, 0.75] x [0, 1] (1)
[0, 1] x [0, 1] x [0.25, 0.25] (1)
[0, 1] x [0, 1] x [0.75, 0.75] (1)
[0, 1] x [0.125, 0.125] x [0, 0.75] (1)
[0, 1] x [0.375, 0.375] x [0, 0.75] (1)
[0, 1] x [0.625, 0.625] x [0, 0.75] (1)
[0, 1] x [0, 0.75] x [0.125, 0.125] (1)
[0, 1] x [0, 0.75] x [0.375, 0.375] (1)
[0, 1] x [0, 0.75] x [0.625, 0.625] (1)
[0.125, 0.125] x [0, 0.75] x [0, 1] (1)
[0.375, 0.375] x [0, 0.75] x [0, 1] (1)
[0.625, 0.625] x [0, 0.75] x [0, 1] (1)
[0.875, 0.875] x [0, 0.75] x [0, 1] (1)
[0, 1] x [0.125, 0.125] x [0, 1] (1)
[0, 1] x [0.375, 0.375] x [0, 1] (1)
[0, 1] x [0.625, 0.625] x [0, 1] (1)
[0, 1] x [0, 0.75] x [0.875, 0.875] (1)
[0, 1] x [0.875, 0.875] x [0, 0.75] (1)
[0, 1] x [0, 1] x [0.125, 0.125] (1)
[0, 1] x [0, 1] x [0.375, 0.375] (1)
[0, 1] x [0, 1] x [0.625, 0.625] (1)
[0.125, 0.125] x [0, 1] x [0, 1] (1)
[0.375, 0.375] x [0, 1] x [0, 1] (1)
[0.625, 0.625] x [0, 1] x [0, 1] (1)
[0.875, 0.875] x [0, 1] x [0, 1] (1)
[0, 1] x [0.875, 0.875] x [0, 1] (1)
[0, 1] x [0, 1] x [0.875, 0.875] (1)
[0.0625, 0.0625] x [0, 0.375] x [0, 0.375] (1)
[0, 0.375] x [0.0625, 0.0625] x [0, 0.375] (1)
[0, 0.375] x [0, 0.375] x [0.0625, 0.0625] (1)
[0.0625, 0.0625] x [0, 0.375] x [0.5, 1] (1)
[0.625, 1] x [0, 0.375] x [0.0625, 0.0625] (1)
[0.625, 1] x [0, 0.375] x [0.1875, 0.1875] (1)
[0.625, 1] x [0, 0.375] x [0.3125, 0.3125] (1)
[0.1875, 0.1875] x [0, 0.375] x [0, 1] (1)
[0.3125, 0.3125] x [0, 0.375] x [0, 1] (1)
[0.125, 0.375] x [0.0625, 0.0625] x [0, 1] (1)
[0, 0.5] x [0.0625, 0.0625] x [0.5, 1] (1)
[0, 0.5] x [0, 0.375] x [0.3125, 0.3125] (1)
[0.125, 0.5] x [0, 0.375] x [0.4375, 0.4375] (1)
[0.0625, 0.0625] x [0.125, 0.375] x [0, 1] (1)
[0, 0.375] x [0.1875, 0.1875] x [0, 1] (1)
[0, 0.375] x [0.3125, 0.3125] x [0, 1] (1)
[0, 0.375] x [0, 0.5] x [0.3125, 0.3125] (1)
[0, 0.375] x [0.125, 0.5] x [0.4375, 0.4375] (1)
[0, 0.5] x [0.125, 0.375] x [0.4375, 0.4375] (1)
[0.125, 0.375] x [0, 0.5] x [0.4375, 0.4375] (1)
[0, 0.375] x [0, 0.5] x [0.5625, 0.5625] (1)
[0.625, 0.625] x [0.0625, 0.0625] x [0, 1] (1)
[0.625, 0.625] x [0.1875, 0.1875] x [0, 1] (1)
[0.625, 0.625] x [0.3125, 0.3125] x [0, 1] (1)
[0.625, 1] x [0, 0.375] x [0.4375, 0.4375] (1)
[0, 1] x [0, 0.375] x [0.6875, 0.6875] (1)
[0.0625, 0.0625] x [0, 0.5] x [0.375, 0.375] (1)
[0.6875, 0.6875] x [0, 0.375] x [0, 1] (1)
[0.8125, 0.8125] x [0, 0.375] x [0, 1] (1)
[0.9375, 0.9375] x [0, 0.375] x [0, 1] (1)
[0, 1] x [0.0625, 0.0625] x [0.625, 1] (1)
[0.625, 1] x [0.0625, 0.0625] x [0, 1] (1)
[0, 1] x [0.1875, 0.1875] x [0.625, 1] (1)
[0.625, 1] x [0.1875, 0.1875] x [0, 1] (1)
[0, 1] x [0.3125, 0.3125] x [0.625, 1] (1)
[0.625, 1] x [0.3125, 0.3125] x [0, 1] (1)
[0, 1] x [0, 0.375] x [0.8125, 0.8125] (1)
[0, 1] x [0, 0.375] x [0.9375, 0.9375] (1)
[0.0625, 0.0625] x [0.625, 1] x [0.625, 1] (1)
[0.1875, 0.1875] x [0.625, 1] x [0.625, 1] (1)
[0.3125, 0.3125] x [0.625, 1] x [0.625, 1] (1)
[0, 0.375] x [0.6875, 0.6875] x [0.625, 1] (1)
[0, 0.375] x [0.8125, 0.8125] x [0.625, 1] (1)
[0, 0.375] x [0.9375, 0.9375] x [0.625, 1] (1)
[0, 0.375] x [0.625, 1] x [0.6875, 0.6875] (1)
[0, 0.375] x [0.625, 1] x [0.8125, 0.8125] (1)
[0, 0.375] x [0.625, 1] x [0.9375, 0.9375] (1)
[0.4375, 0.4375] x [0, 0.5] x [0.625, 1] (1)
[0.5625, 0.5625] x [0, 0.5] x [0.625, 1] (1)
[0.6875, 0.6875] x [0, 0.5] x [0.625, 1] (1)
[0, 0.75] x [0, 0.5] x [0.6875, 0.6875] (1)
[0.0625, 0.0625] x [0, 0.5] x [0.5, 1] (1)
[0, 0.75] x [0.4375, 0.4375] x [0.625, 1] (1)
[0, 0.75] x [0, 0.5] x [0.8125, 0.8125] (1)
[0, 0.75] x [0, 0.5] x [0.9375, 0.9375] (1)
[0.4375, 0.4375] x [0, 0.375] x [0.125, 1] (1)
[0.125, 0.5] x [0.0625, 0.0625] x [0.125, 1] (1)
[0, 0.5] x [0.0625, 0.0625] x [0.125, 0.375] (1)
[0, 0.5] x [0.1875, 0.1875] x [0.125, 1] (1)
[0, 0.5] x [0.3125, 0.3125] x [0.125, 1] (1)
[0, 0.5] x [0, 0.375] x [0.1875, 0.1875] (1)
[0.0625, 0.0625] x [0.125, 0.5] x [0.125, 1] (1)
[0.0625, 0.0625] x [0, 0.5] x [0.125, 0.375] (1)
[0.0625, 0.0625] x [0, 0.5] x [0.5, 0.5] (1)
[0.1875, 0.1875] x [0, 0.5] x [0.125, 1] (1)
[0.3125, 0.3125] x [0, 0.5] x [0.125, 1] (1)
[0, 0.375] x [0.4375, 0.4375] x [0.125, 1] (1)
[0, 0.375] x [0, 0.5] x [0.1875, 0.1875] (1)
[0, 1] x [0.0625, 0.0625] x [0.5, 1] (1)
[0, 1] x [0.1875, 0.1875] x [0.5, 1] (1)
[0, 1] x [0.3125, 0.3125] x [0.5, 1] (1)
[0, 1] x [0, 0.375] x [0.5625, 0.5625] (1)
[0.4375, 0.4375] x [0, 0.5] x [0.5, 1] (1)
[0.5625, 0.5625] x [0, 0.5] x [0.5, 1] (1)
[0, 0.625] x [0.4375, 0.4375] x [0.5, 1] (1)
[0, 0.625] x [0, 0.5] x [0.5625, 0.5625] (1)
```
which eventually leads to a deref at index -1 in the element cache,
```
glob_knot_v_:  0.0625 0.125 0.1875 0.25 0.3125 0.375 0.4375 0.5 0.625 0.6875 0.75 0.8125 0.875 0.9375 1 

#     idx (= size_t(-1), len(glob_knot_v), v
ish j 18446744073709551615 15 0.00433949
```

**Comment by VikingScientist** [comment]
Identified as a pure LR-spline bug. Fix is underway...

**Comment by VikingScientist** [comment]
Fix available in [LRsplines 1.4.1](https://github.com/VikingScientist/LRsplines/releases/tag/PPA-release1.4.1)

**Comment by VikingScientist** [comment]
Fixed in #201
